<a href="https://colab.research.google.com/github/muhammadabdurrehmanmaqsood/flyrank-ml-internship-abdurrehman/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/muhammadabdurrehmanmaqsood/flyrank-ml-internship-abdurrehman/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [9]:
# Selected Lane: Lane 2: Refresh / Content Opportunity Scoring

# Why: The goal is to build a decision-support tool that prioritizes our content team's backlog.
# Instead of relying on guesswork to figure out which pages are losing traffic, we can use
# observable search signals to identify pages that are currently declining but still have
# enough baseline demand to be worth fixing. This translates a raw data problem into an
# actionable, ranked queue for human reviewers.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [10]:
# The Decision: Which specific, currently declining pages should the content team review first today?

# The Action: A human reviewer opens the top-ranked page in the queue
# and decides whether to update the content, consolidate it with another page,
# or leave it alone based on the provided reason codes.

# The Cost of a Wrong Call: If the model ranks a low-value, strictly seasonal,
# or permanently dead page at the top of the queue, a reviewer wastes valuable hours
# working on a page that has no potential to drive meaningful traffic,
# ignoring higher-value opportunities.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [11]:
import pandas as pd

# Load the dataset directly from the raw GitHub URL
url = 'https://raw.githubusercontent.com/muhammadabdurrehmanmaqsood/flyrank-ml-internship-abdurrehman/main/data/raw/content_refresh_anonymized.csv'
df = pd.read_csv(url)

# 1. Total pages in the dataset
total_pages = len(df)

# 2. How many pages are currently flagged as declining?
declining_pages = len(df[df['trend_direction'] == 'down'])

# 3. Of those declining pages, how many have meaningful traffic at risk? (e.g., > 500 impressions)
high_value_declining = len(df[(df['trend_direction'] == 'down') & (df['impressions_90d'] > 500)])

print(f"Total pages in starter dataset: {total_pages}")
print(f"Pages showing a downward trend: {declining_pages}")
print(f"Declining pages with >500 recent impressions (our primary targets): {high_value_declining}")

Total pages in starter dataset: 30000
Pages showing a downward trend: 16262
Declining pages with >500 recent impressions (our primary targets): 9956


To justify this lane, I checked the starter dataset (data/raw/content_refresh_anonymized.csv):



1.   There are **30000** total pages in the starter slice.
2.   Of those, **16262** pages are actively showing a downward trend.
3.   More importantly, **9956** of those declining pages have generated over **500** impressions in the last **90** days. This proves there is a substantial, high-value backlog that requires prioritization, rather than just a handful of dead pages.



## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [12]:
# What I can claim: I can claim that we can identify and rank pages with a high
# probability of matching past decline patterns based purely on observable search signals.
# We can surface the most urgent candidates for human review.

# What I cannot claim: I cannot guarantee that editing a recommended page will cause a
# recovery in traffic. Proving that a refresh caused a recovery requires an explicit
# experiment or a causal design, which this purely observational dataset does not support.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.